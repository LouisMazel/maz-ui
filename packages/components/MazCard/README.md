# MazCard

## Props

<!-- @vuese:MazCard:props:start -->

| Name          | Description                                                                         | Type                | Required | Default |
| ------------- | ----------------------------------------------------------------------------------- | ------------------- | -------- | ------- |
| images        | Images displayed                                                                    | `Array`             | `false`  | null    |
| variant       | Card variant: Must be `column | row | row-reverse | column-reverse`                 | `String`            | `false`  | column  |
| href          | Make card a link (footer area excluded)                                             | `String`            | `false`  | null    |
| hrefTarget    | Target option of link: Muse be one of `_blank | _self | _parent | _top | framename` | `String`            | `false`  | \_self  |
| width         | Card width                                                                          | `String` / `Number` | `false`  | null    |
| maxWidth      | Card max-width                                                                      | `String` / `Number` | `false`  | 400     |
| footerAlign   | Footer text alignment: `right | left`                                               | `String`            | `false`  | right   |
| galleryWidth  | Gallery image width                                                                 | `String` / `Number` | `false`  | 200     |
| galleryHeight | Gallery image height                                                                | `String` / `Number` | `false`  | 150     |
| zoom          | Enable "zoom" image on click (can't be used when the card has a link)               | `Boolean`           | `false`  | false   |
| elevation     | Set elevation to card (box-shadow)                                                  | `Boolean`           | `false`  | true    |
| radius        | Set radius to card (box-shadow)                                                     | `Boolean`           | `false`  | true    |
| bordered      | Set border to card                                                                  | `Boolean`           | `false`  | false   |

<!-- @vuese:MazCard:props:end -->

## Slots

<!-- @vuese:MazCard:slots:start -->

| Name     | Description                           | Default Slot Content |
| -------- | ------------------------------------- | -------------------- |
| actions  | Can't be used when the card is linked | -                    |
| default  | -                                     | -                    |
| title    | -                                     | -                    |
| subtitle | -                                     | -                    |
| content  | -                                     | -                    |
| footer   | -                                     | -                    |

<!-- @vuese:MazCard:slots:end -->
