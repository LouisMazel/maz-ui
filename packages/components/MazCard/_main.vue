<template>
  <div
    class="maz-card maz-base-component maz-flex maz-direction-column"
    :class="{
      'maz-card--linked': href,
      'maz-elevation': elevation,
      'maz-border-radius': radius,
      'maz-border maz-border-solid maz-border-color': bordered
    }"
    :style="[cardStyle]"
  >
    <component
      :is="href ? 'a' : 'div'"
      class="maz-card__wrapper maz-flex maz-flex-1"
      :class="[`maz-direction-${variant}`, { 'maz-card__link': href }]"
      :href="href"
      :target="href ? hrefTarget : null"
    >
      <div class="maz-card__gallery__wrapper maz-flex">
        <MazGallery
          v-if="images"
          :radius="radius"
          :width="_galleryWidth"
          :height="galleryHeight"
          :images="images"
          :zoom="zoom"
          :no-width="isColumnVariant"
          :no-height="!isColumnVariant && haveSomeContent"
          class="maz-card__gallery maz-flex-fixed maz-flex-1"
        />
        <div
          v-if="$slots['actions']"
          class="maz-card__gallery__actions maz-p-2 maz-text-left"
        >
          <!-- Can't be used when the card is linked -->
          <slot name="actions" />
        </div>
      </div>
      <!-- {{ $slots['title'] }} -->
      <div
        v-if="haveSomeContent"
        class="maz-flex-1"
      >
        <div class="maz-card__content__wrapper maz-p-3">
          <slot>
            <div
              v-if="$slots['title']"
              class="maz-card__title"
            >
              <slot name="title" />
            </div>
            <div
              v-if="$slots['subtitle']"
              class="maz-card__subtitle maz-text-muted"
            >
              <slot name="subtitle" />
            </div>
            <div
              v-if="$slots['content']"
              class="maz-card__content maz-pt-2"
            >
              <slot name="content" />
            </div>
          </slot>
        </div>
      </div>
    </component>
    <div
      v-if="$slots['footer']"
      class="maz-card__footer maz-p-3"
      :class="[{ 'maz-border-top maz-border-color maz-border-top-solid': isColumnVariant && haveSomeContent }, `maz-text-${footerAlign}`]"
    >
      <slot name="footer" />
    </div>
  </div>
</template>



<script>
export default {
  name: 'MazCard',
  props: {
    // Images displayed
    images: { type: Array, default: null },
    // Card variant: Must be `column | row | row-reverse | column-reverse`
    variant: { type: String, default: 'column' },
    // Make card a link (footer area excluded)
    href: { type: String, default: null },
    // Target option of link: Muse be one of `_blank | _self | _parent | _top | framename`
    hrefTarget: { type: String, default: '_self' },
    // Card width
    width: { type: [String, Number], default: null },
    // Card max-width
    maxWidth: { type: [String, Number], default: 400 },
    // Footer text alignment: `right | left`
    footerAlign: { type: String, default: 'right' },
    // Gallery image width
    galleryWidth: { type: [String, Number], default: 200 },
    // Gallery image height
    galleryHeight: { type: [String, Number], default: 150 },
    // Enable "zoom" image on click (can't be used when the card has a link)
    zoom: { type: Boolean, default: false },
    // Set elevation to card (box-shadow)
    elevation: { type: Boolean, default: true },
    // Set radius to card (box-shadow)
    radius: { type: Boolean, default: true },
    // Set border to card
    bordered: { type: Boolean, default: false }
  },
  computed: {
    isColumnVariant () {
      const { variant } = this
      return ['column', 'column-reverse'].includes(variant)
    },
    cardStyle () {
      const { width, maxWidth } = this
      return {
        width: !width ? null : Number.isInteger(width) ? `${width}px` : width,
        minWidth: !width ? null : Number.isInteger(width) ? `${width}px` : width,
        maxWidth: width ? null : Number.isInteger(maxWidth) ? `${maxWidth}px` : maxWidth
      }
    },
    haveSomeContent () {
      const { $slots } = this
      const slots = ['default', 'title', 'subtitle', 'content']
      const response = Object.keys($slots).some((val) => slots.includes(val))
      return response
    },
    _galleryWidth () {
      const { galleryWidth, haveSomeContent } = this
      return haveSomeContent ? galleryWidth : '100%'
    }
  }
}
</script>

