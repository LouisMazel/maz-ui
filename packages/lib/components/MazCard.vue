<template>
  <div
    class="m-card"
    :class="[
      {
        'm-card--linked': href,
        'm-card--no-scale': !scale,
        'maz-elevation': elevation,
        'maz-overflow-hidden': overflowHidden || !isOpen,
        'maz-rounded': radius,
        'maz-border maz-border-solid maz-border-color-light': bordered,
      },
    ]"
  >
    <Component
      :is="collapsable ? 'button' : 'div'"
      v-if="$slots['header'] || header || collapsable"
      class="m-card__header maz-border-b maz-border-solid"
      :class="[
        isOpen
          ? 'maz-rounded-t maz-border-color-light dark:maz-bg-color-lighter'
          : 'maz-border-transparent',
        { '--is-collapsable': collapsable },
        { 'maz-justify-end': (!$slots['header'] || !header) && collapsable },
        { 'maz-justify-between': $slots['header'] || header },
      ]"
      tabindex="-1"
      @click.stop="collapsable ? (isOpen = !isOpen) : undefined"
    >
      <slot v-if="$slots['header'] || header" name="header">{{ header }}</slot>

      <MazBtn
        v-if="collapsable"
        color="transparent"
        class="maz-ml-2 maz-text-sm"
        size="xs"
        @click.stop="isOpen = !isOpen"
      >
        <ChevronDownIcon
          :class="{ '--is-open': isOpen }"
          class="m-card__collapse-icon maz-text-xl"
        />
      </MazBtn>
    </Component>
    <Component
      :is="wrapperData.is"
      v-bind="wrapperData"
      class="m-card__wrapper"
      :class="[`m-card__wrapper--${orientation}`, { 'm-card__link': href }]"
    >
      <div v-if="images" class="m-card__gallery__wrapper">
        <MazGallery
          :no-radius="!radius"
          :width="galleryWidthComputed"
          :height="galleryHeight"
          :images-shown-count="imagesShowCount"
          :images="images"
          :no-zoom="!zoom"
          :no-width="isColumnVariant"
          :no-height="!isColumnVariant && haveSomeContent"
          :no-remaining="noRemaining"
          class="m-card__gallery maz-flex-1"
        />
      </div>
      <div class="maz-min-w-0 maz-flex-1">
        <Component :is="collapsable ? MazTransitionExpand : 'div'">
          <div
            v-show="isOpen"
            :class="[wrapperClass, { 'maz-p-4': !noPadding && !collapsable }]"
            class="m-card__content__wrapper"
          >
            <slot>
              <div v-if="$slots['title']" class="m-card__title">
                <slot name="title"></slot>
              </div>
              <div v-if="$slots['subtitle']" class="m-card__subtitle">
                <slot name="subtitle"></slot>
              </div>
              <div v-if="$slots['content']" class="m-card__content">
                <slot name="content"></slot>
              </div>
            </slot>
          </div>
        </Component>
      </div>
    </Component>
    <div
      v-if="$slots['footer']"
      class="m-card__footer maz-overflow-x-auto maz-p-3"
      :class="[
        {
          'maz-border-t-2 maz-border-color-light dark:maz-bg-color-lighter':
            isColumnVariant && haveSomeContent,
        },
        footerAlignClass,
      ]"
    >
      <slot name="footer"></slot>
    </div>
    <div v-if="$slots['actions']" class="m-card__actions maz-flex maz-p-2">
      <!-- @slot action of gallery -->
      <slot name="actions"></slot>
    </div>
  </div>
</template>

<script lang="ts" setup>
  import { computed, useSlots, ref, watch, defineAsyncComponent, type HTMLAttributes } from 'vue'
  import type { MazGalleryImage } from './types'
  import { type RouterLinkProps } from 'vue-router'
  export type { MazGalleryImage } from './types'

  const MazBtn = defineAsyncComponent(() => import('./MazBtn.vue'))
  const MazGallery = defineAsyncComponent(() => import('./MazGallery.vue'))
  const MazTransitionExpand = defineAsyncComponent(() => import('./MazTransitionExpand.vue'))
  const ChevronDownIcon = defineAsyncComponent(() => import('./../icons/chevron-down.svg'))

  export type Props = {
    /** Images displayed */
    images?: MazGalleryImage[]
    /** Card variant: Must be `column | row | row-reverse | column-reverse` */
    orientation?: 'column' | 'row' | 'row-reverse' | 'column-reverse'
    /** Make card a link (footer area excluded) */
    href?: string
    /** Make card a link with a router-link (footer area excluded) */
    to?: RouterLinkProps['to']
    /** Target option of link: Muse be one of `_blank | _self | _parent | _top | framename` */
    hrefTarget?: '_blank' | '_self' | '_parent' | '_top' | string
    /** Footer text alignment: `right | left` */
    footerAlign?: 'right' | 'left'
    /** Gallery image width */
    galleryWidth?: string | number
    /** Gallery image height */
    galleryHeight?: string | number
    /** Enable "zoom" image on click (can't be used when the card has a link) */
    zoom?: boolean
    /** Set elevation to card (box-shadow) */
    elevation?: boolean
    /** Set radius to card */
    radius?: boolean
    /** Set border to card */
    bordered?: boolean
    /** Number of images shown in the gallery */
    imagesShowCount?: number
    /** Remove transparent layer with the remain count (ex: +2) */
    noRemaining?: boolean
    /** scale animation on hover (only linked cards) */
    scale?: boolean
    /** add classes to wrapper */
    wrapperClass?: HTMLAttributes['class']
    /** Remove padding from content wrapper */
    noPadding?: boolean
    /** Hide overflow */
    overflowHidden?: boolean
    /** Card can be open and close */
    collapsable?: boolean
    /** Card is open by default if `true` */
    collapseOpen?: boolean
    /** Title of the card in header */
    header: string
  }

  const props = withDefaults(defineProps<Props>(), {
    images: undefined,
    orientation: 'column',
    href: undefined,
    to: undefined,
    hrefTarget: '_self',
    footerAlign: 'right',
    galleryWidth: 200,
    galleryHeight: 150,
    elevation: true,
    radius: true,
    imagesShowCount: 3,
    noRemaining: true,
    scale: true,
    wrapperClass: undefined,
  })

  const slots = useSlots()

  const isOpen = ref(props.collapsable ? props.collapseOpen : true)

  watch(
    () => props.collapseOpen,
    (value) => {
      if (props.collapsable) isOpen.value = value
    },
  )

  const wrapperData = computed(() => {
    return {
      is: props.href ? 'a' : props.to ? 'router-link' : 'div',
      ...(props.href && { href: props.href }),
      ...(props.to && { to: props.to }),
      target: props.hrefTarget,
    }
  })

  const isColumnVariant = computed(() => ['column', 'column-reverse'].includes(props.orientation))

  const haveSomeContent = computed(() => {
    const supportedSlots = new Set(['default', 'title', 'subtitle', 'content'])
    return Object.keys(slots).some((val) => supportedSlots.has(val))
  })

  const galleryWidthComputed = computed(() => (haveSomeContent.value ? props.galleryWidth : '100%'))

  const footerAlignClass = computed(() =>
    props.footerAlign === 'right' ? 'maz-text-right' : 'maz-text-left',
  )
</script>

<style lang="postcss" scoped>
  .m-card {
    @apply maz-relative maz-inline-flex maz-max-h-full maz-flex-col maz-bg-color dark:maz-border dark:maz-border-color-lighter;

    &__header {
      @apply maz-flex maz-items-center maz-px-4 maz-py-3 maz-transition-colors maz-duration-200;

      &.--is-collapsable {
        @apply hover:maz-bg-color-light;
      }
    }

    &__collapse-icon {
      @apply maz-rotate-0 maz-transition-transform maz-duration-200;

      &.--is-open {
        transform: rotate(180deg);
      }
    }

    &__wrapper {
      @apply maz-flex maz-flex-1;

      &--row {
        @apply maz-flex-row;
      }

      &--row-reverse {
        @apply maz-flex-row-reverse;
      }

      &--column {
        @apply maz-flex-col;
      }

      &--column-reverse {
        @apply maz-flex-col-reverse;
      }
    }

    &--linked {
      transition: all 300ms ease-in-out;
      transform: scale(1);

      &:hover:not(.m-card--no-scale) {
        @apply maz-z-1;

        transform: scale(1.02);
      }

      & .m-card__wrapper {
        @apply maz-cursor-pointer maz-no-underline;

        &:hover {
          @apply maz-no-underline;
        }
      }
    }

    &__content__wrapper {
      @apply maz-relative maz-max-w-full;
    }

    &__title,
    &__title > * {
      font-size: 1.2em;

      @apply maz-text-normal;
    }

    &__subtitle,
    &__subtitle > * {
      font-size: 1.1em;

      @apply maz-text-muted;
    }

    &__gallery__wrapper {
      @apply maz-relative maz-flex maz-overflow-hidden;
    }

    &__actions {
      @apply maz-absolute maz-left-0 maz-top-0 maz-z-2;

      & > *:not(:last-child) {
        @apply maz-mr-2;
      }
    }
  }
</style>
