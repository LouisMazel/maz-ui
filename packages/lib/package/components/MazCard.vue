<template>
  <div
    class="m-card"
    :class="[
      {
        'm-card--linked': href,
        'm-card--no-scale': !scale,
        'maz-elevation': elevation,
        'maz-overflow-hidden': overflowHidden,
        'maz-rounded-xl': radius,
        'maz-border-2 maz-border-solid maz-border-color-lighter': bordered,
      },
    ]"
  >
    <Component
      :is="collapsable ? 'button' : 'div'"
      v-if="$slots['header'] || collapsable"
      class="m-card__header maz-rounded-t-xl maz-border-b-2"
      :class="[
        isOpen ? 'maz-border-color-lighter' : 'maz-border-transparent',
        { '--is-collapsable': collapsable },
      ]"
      tabindex="-1"
      @click.stop="collapsable ? (isOpen = !isOpen) : undefined"
    >
      <slot name="header" />

      <MazBtn
        v-if="collapsable"
        color="transparent"
        class="maz-ml-2 maz-text-sm"
        size="xs"
        @click.stop="isOpen = !isOpen"
      >
        <MazIcon
          :src="ChevronDownIcon"
          :class="{ '--is-open': isOpen }"
          class="m-card__collapse-icon maz-h-6"
        />
      </MazBtn>
    </Component>
    <Component
      :is="href ? 'a' : 'div'"
      class="m-card__wrapper"
      :class="[`m-card__wrapper--${orientation}`, { 'm-card__link': href }]"
      :href="href"
      :target="href ? hrefTarget : null"
    >
      <div v-if="images" class="m-card__gallery__wrapper">
        <MazGallery
          :radius="radius"
          :width="galleryWidthComputed"
          :height="galleryHeight"
          :images-shown-count="imagesShowCount"
          :images="images"
          :zoom="zoom"
          :no-width="isColumnVariant"
          :no-height="!isColumnVariant && haveSomeContent"
          :no-remaining="noRemaining"
          class="m-card__gallery maz-flex-1"
        />
      </div>
      <div class="maz-min-w-0 maz-flex-1">
        <Component :is="collapsable ? MazTransitionExpand : 'div'">
          <div
            v-show="isOpen"
            :class="[wrapperClass, { 'maz-p-4': !noPadding && !collapsable }]"
            class="m-card__content__wrapper"
          >
            <slot>
              <div v-if="$slots['title']" class="m-card__title">
                <slot name="title" />
              </div>
              <div v-if="$slots['subtitle']" class="m-card__subtitle">
                <slot name="subtitle" />
              </div>
              <div v-if="$slots['content']" class="m-card__content">
                <slot name="content" />
              </div>
            </slot>
          </div>
        </Component>
      </div>
    </Component>
    <div
      v-if="$slots['footer']"
      class="m-card__footer maz-overflow-x-auto maz-p-3"
      :class="[
        {
          'maz-border-t-2 maz-border-color-lighter':
            isColumnVariant && haveSomeContent,
        },
        footerAlignClass,
      ]"
    >
      <slot name="footer" />
    </div>
    <div v-if="$slots['actions']" class="m-card__actions maz-flex maz-p-2">
      <!-- Slot above the gallery -->
      <slot name="actions" />
    </div>
  </div>
</template>

<script lang="ts">
  export type { MazGalleryImage } from './types'
</script>

<script lang="ts" setup>
  import { computed, useSlots, PropType, ref, watch } from 'vue'
  import MazGallery from './MazGallery.vue'
  import MazBtn from './MazBtn.vue'
  import MazTransitionExpand from './MazTransitionExpand.vue'
  import MazIcon from './MazIcon.vue'
  import { MazGalleryImage } from './types'
  import ChevronDownIcon from './../icons/chevron-down.svg'

  const props = defineProps({
    // Images displayed
    images: {
      type: Array as PropType<MazGalleryImage[]>,
      default: undefined,
    },
    // Card variant: Must be `column | row | row-reverse | column-reverse`
    orientation: {
      type: String,
      default: 'column',
      validator: (value: string) => {
        return ['column', 'row', 'row-reverse', 'column-reverse'].includes(
          value,
        )
      },
    },
    // Make card a link (footer area excluded)
    href: { type: String, default: undefined },
    // Target option of link: Muse be one of `_blank | _self | _parent | _top | framename`
    hrefTarget: { type: String, default: '_self' },
    // Footer text alignment: `right | left`
    footerAlign: { type: String, default: 'right' },
    // Gallery image width
    galleryWidth: { type: [String, Number], default: 200 },
    // Gallery image height
    galleryHeight: { type: [String, Number], default: 150 },
    // Enable "zoom" image on click (can't be used when the card has a link)
    zoom: { type: Boolean, default: false },
    // Set elevation to card (box-shadow)
    elevation: { type: Boolean, default: true },
    // Set radius to card (box-shadow)
    radius: { type: Boolean, default: true },
    // Set border to card
    bordered: { type: Boolean, default: false },
    // Number of images shown in the gallery
    imagesShowCount: { type: Number, default: 3 },
    // Remove transparent layer with the remain count (ex: +2)
    noRemaining: { type: Boolean, default: true },
    // scale animation on hover (only linked cards)
    scale: { type: Boolean, default: true },
    wrapperClass: { type: String, default: undefined },
    noPadding: { type: Boolean, default: false },
    overflowHidden: { type: Boolean, default: false },
    collapsable: { type: Boolean, default: false },
    collapseOpen: { type: Boolean, default: false },
  })

  const slots = useSlots()

  const isOpen = ref(props.collapsable ? props.collapseOpen : true)

  watch(
    () => props.collapseOpen,
    (value) => {
      if (props.collapsable) isOpen.value = value
    },
  )

  const isColumnVariant = computed(() =>
    ['column', 'column-reverse'].includes(props.orientation),
  )

  const haveSomeContent = computed(() => {
    const supportedSlots = ['default', 'title', 'subtitle', 'content']
    const response = Object.keys(slots).some((val) =>
      supportedSlots.includes(val),
    )
    return response
  })

  const galleryWidthComputed = computed(() =>
    haveSomeContent.value ? props.galleryWidth : '100%',
  )

  const footerAlignClass = computed(() =>
    props.footerAlign === 'right' ? 'maz-text-right' : 'maz-text-left',
  )
</script>

<style lang="postcss" scoped>
  .m-card {
    @apply maz-relative maz-inline-flex maz-max-h-full maz-flex-col maz-bg-color;

    &__header {
      @apply maz-flex maz-items-center maz-justify-between maz-py-3 maz-px-4 maz-transition-colors maz-delay-200;

      &.--is-collapsable {
        @apply hover:maz-bg-color-lighter;
      }
    }

    &__collapse-icon {
      @apply maz-rotate-0 maz-transition-transform maz-duration-200;

      &.--is-open {
        transform: rotate(180deg);
      }
    }

    &__wrapper {
      @apply maz-flex maz-flex-1;

      &--row {
        @apply maz-flex-row;
      }

      &--row-reverse {
        @apply maz-flex-row-reverse;
      }

      &--column {
        @apply maz-flex-col;
      }

      &--column-reverse {
        @apply maz-flex-col-reverse;
      }
    }

    &--linked {
      transition: all 300ms ease-in-out;
      transform: scale(1);

      &:hover:not(.m-card--no-scale) {
        @apply maz-z-1;

        transform: scale(1.02);
      }

      & .m-card__wrapper {
        @apply maz-cursor-pointer maz-no-underline;

        &:hover {
          @apply maz-no-underline;
        }
      }
    }

    &__content__wrapper {
      @apply maz-relative maz-max-w-full;
    }

    &__title,
    &__title > * {
      font-size: 1.2em;

      @apply maz-text-normal-text;
    }

    &__subtitle,
    &__subtitle > * {
      font-size: 1.1em;

      @apply maz-text-muted;
    }

    &__gallery__wrapper {
      @apply maz-relative maz-flex maz-overflow-hidden;
    }

    &__actions {
      @apply maz-absolute maz-top-0 maz-left-0 maz-z-2;

      & > *:not(:last-child) {
        @apply maz-mr-2;
      }
    }
  }

  html.dark {
    & .m-card {
      @apply maz-bg-color-light;
    }
  }
</style>
