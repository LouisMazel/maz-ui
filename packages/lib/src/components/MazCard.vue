<script lang="ts" setup>
import type { HTMLAttributes } from 'vue'
import type { RouterLinkProps } from 'vue-router'
import type { MazGalleryImage } from './types'
import { computed, defineAsyncComponent, ref, useSlots, watch } from 'vue'

const props = withDefaults(defineProps<MazCardProps>(), {
  images: undefined,
  orientation: 'column',
  href: undefined,
  to: undefined,
  hrefTarget: '_self',
  footerAlign: 'right',
  galleryWidth: 200,
  galleryHeight: 150,
  elevation: true,
  radius: true,
  imagesShowCount: 3,
  noRemaining: true,
  scale: true,
  wrapperClass: undefined,
  header: undefined,
})

const MazBtn = defineAsyncComponent(() => import('./MazBtn.vue'))
const MazGallery = defineAsyncComponent(() => import('./MazGallery.vue'))
const MazTransitionExpand = defineAsyncComponent(() => import('./MazTransitionExpand.vue'))

const ChevronDownIcon = defineAsyncComponent(() => import('../../icons/chevron-down.svg'))

export interface MazCardProps {
  /** Images displayed */
  images?: MazGalleryImage[]
  /** Card variant: Must be `column | row | row-reverse | column-reverse` */
  orientation?: 'column' | 'row' | 'row-reverse' | 'column-reverse'
  /** Make card a link (footer area excluded) */
  href?: string
  /** Make card a link with a router-link (footer area excluded) */
  to?: RouterLinkProps['to']
  /** Target option of link: Muse be one of `_blank | _self | _parent | _top | framename` */
  hrefTarget?: '_blank' | '_self' | '_parent' | '_top' | string
  /** Footer text alignment: `right | left` */
  footerAlign?: 'right' | 'left'
  /** Gallery image width */
  galleryWidth?: string | number
  /** Gallery image height */
  galleryHeight?: string | number
  /** Enable "zoom" image on click (can't be used when the card has a link) */
  zoom?: boolean
  /** Set elevation to card (box-shadow) */
  elevation?: boolean
  /** Set radius to card */
  radius?: boolean
  /** Set border to card (in dark mode, the card is always bordered) */
  bordered?: boolean
  /** Number of images shown in the gallery */
  imagesShowCount?: number
  /** Remove transparent layer with the remain count (ex: +2) */
  noRemaining?: boolean
  /** scale animation on hover (only linked cards) */
  scale?: boolean
  /** add classes to wrapper */
  wrapperClass?: HTMLAttributes['class']
  /** Remove padding from content wrapper */
  noPadding?: boolean
  /** Hide overflow */
  overflowHidden?: boolean
  /**
   * @deprecated Use `collapsible` instead
   */
  collapsable?: boolean
  /**
   * Card can be open and close
   */
  collapsible?: boolean
  /** Card is open by default if `true` */
  collapseOpen?: boolean
  /** Title of the card in header */
  header?: string
  /** The card will be displayed in full width */
  block?: boolean
}

const slots = useSlots()

const isCollapsible = computed(() => props.collapsible || props.collapsable)

const isOpen = ref(isCollapsible.value ? props.collapseOpen : true)

const isLinked = computed(() => props.href || props.to)

watch(
  () => props.collapseOpen,
  (value) => {
    if (isCollapsible.value)
      isOpen.value = value
  },
)

const wrapperData = computed(() => {
  return {
    is: props.href ? 'a' : props.to ? 'router-link' : 'div',
    ...(props.href && { href: props.href }),
    ...(props.to && { to: props.to }),
    target: props.hrefTarget,
  }
})

const isColumnVariant = computed(() => ['column', 'column-reverse'].includes(props.orientation))

const haveSomeContent = computed(() => {
  const supportedSlots = new Set(['default', 'title', 'subtitle', 'content'])
  return Object.keys(slots).some(val => supportedSlots.has(val))
})

const galleryWidthComputed = computed(() => (haveSomeContent.value ? props.galleryWidth : '100%'))

const footerAlignClass = computed(() =>
  props.footerAlign === 'right' ? 'maz-text-end' : 'maz-text-start',
)
</script>

<template>
  <div
    class="m-card m-reset-css"
    :class="[
      {
        'm-card--linked': isLinked,
        'm-card--no-scale': !scale,
        'maz-elevation': elevation,
        '--block': block,
        'maz-overflow-hidden': overflowHidden || !isOpen,
        'maz-rounded': radius,
        'maz-border maz-border-solid maz-border-color-light': bordered,
      },
    ]"
  >
    <Component
      :is="isCollapsible ? 'button' : 'div'"
      v-if="$slots.header || header || isCollapsible"
      class="m-card__header maz-border-b maz-border-solid"
      :class="[
        isOpen ? 'maz-rounded-t maz-border-color-light' : 'maz-border-transparent',
        { '--is-collapsible': isCollapsible },
        { 'maz-justify-end': (!$slots.header || !header) && isCollapsible },
        { 'maz-justify-between': $slots.header || header },
      ]"
      tabindex="-1"
      @click.stop="isCollapsible ? (isOpen = !isOpen) : undefined"
    >
      <slot v-if="$slots.header || header" name="header">
        {{ header }}
      </slot>

      <MazBtn
        v-if="isCollapsible"
        color="transparent"
        class="maz-ml-2 maz-text-sm"
        size="xs"
        @click.stop="isOpen = !isOpen"
      >
        <ChevronDownIcon
          :class="{ '--is-open': isOpen }"
          class="m-card__collapse-icon maz-text-xl"
        />
      </MazBtn>
    </Component>
    <Component
      :is="wrapperData.is"
      v-bind="wrapperData"
      class="m-card__wrapper"
      :class="[`m-card__wrapper--${orientation}`, { 'm-card__link': isLinked }]"
    >
      <div v-if="images" class="m-card__gallery__wrapper">
        <MazGallery
          :no-radius="!radius"
          :width="galleryWidthComputed"
          :height="galleryHeight"
          :images-shown-count="imagesShowCount"
          :images="images"
          :no-zoom="!zoom"
          :no-width="isColumnVariant"
          :no-height="!isColumnVariant && haveSomeContent"
          :no-remaining="noRemaining"
          class="m-card__gallery maz-flex-1"
        />
      </div>
      <div class="maz-min-w-0 maz-flex-1">
        <Component :is="isCollapsible ? MazTransitionExpand : 'div'" class="maz-h-full">
          <div
            v-show="isOpen"
            :class="[wrapperClass, { 'maz-p-4': !noPadding && !isCollapsible }]"
            class="m-card__content__wrapper maz-h-full"
          >
            <slot>
              <div v-if="$slots.title" class="m-card__title">
                <slot name="title" />
              </div>
              <div v-if="$slots.subtitle" class="m-card__subtitle">
                <slot name="subtitle" />
              </div>
              <div v-if="$slots.content" class="m-card__content">
                <slot name="content" />
              </div>
            </slot>
          </div>
        </Component>
      </div>
    </Component>
    <div
      v-if="$slots.footer"
      class="m-card__footer maz-overflow-x-auto maz-p-3"
      :class="[
        {
          'maz-border-t maz-border-color-light': isColumnVariant && haveSomeContent,
        },
        footerAlignClass,
      ]"
    >
      <slot name="footer" />
    </div>
    <div v-if="$slots.actions" class="m-card__actions maz-flex maz-p-2">
      <!-- @slot action of gallery -->
      <slot name="actions" />
    </div>
  </div>
</template>

<style lang="postcss" scoped>
  .m-card {
  @apply maz-relative maz-inline-flex maz-max-h-full maz-flex-col maz-bg-color dark:maz-border dark:maz-border-color-light;

  &.--block {
    @apply maz-w-full;
  }

  &__header {
    @apply maz-flex maz-items-center maz-px-4 maz-py-3 maz-transition-colors maz-duration-200;

    &.--is-collapsible {
      @apply hover:maz-bg-color-light;
    }
  }

  &__collapse-icon {
    @apply maz-rotate-0 maz-transition-transform maz-duration-200;

    &.--is-open {
      transform: rotate(180deg);
    }
  }

  &__wrapper {
    @apply maz-flex maz-flex-1;

    &--row {
      @apply maz-flex-row;
    }

    &--row-reverse {
      @apply maz-flex-row-reverse;
    }

    &--column {
      @apply maz-flex-col;
    }

    &--column-reverse {
      @apply maz-flex-col-reverse;
    }
  }

  &--linked {
    transition: all 300ms ease-in-out;
    transform: scale(1);

    &:hover:not(.m-card--no-scale) {
      @apply maz-z-1;

      transform: scale(1.02);
    }

    & .m-card__wrapper {
      @apply maz-cursor-pointer maz-no-underline;

      &:hover {
        @apply maz-no-underline;
      }
    }
  }

  &__content__wrapper {
    @apply maz-relative maz-max-w-full;
  }

  &__title,
  &__title > * {
    font-size: 1.2em;

    @apply maz-text-normal;
  }

  &__subtitle,
  &__subtitle > * {
    font-size: 1.1em;

    @apply maz-text-muted;
  }

  &__gallery__wrapper {
    @apply maz-relative maz-flex maz-overflow-hidden;
  }

  &__actions {
    @apply maz-absolute maz-left-0 maz-top-0 maz-z-2;

    & > *:not(:last-child) {
      @apply maz-mr-2;
    }
  }
}
</style>
