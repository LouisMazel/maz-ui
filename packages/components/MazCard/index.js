import MazCard from './_main.vue'

MazCard.install = (Vue) => {
  Vue.component(MazCard.name, MazCard)
}

export default MazCard
