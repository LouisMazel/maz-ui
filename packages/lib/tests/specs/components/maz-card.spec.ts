import { mount } from '@vue/test-utils'
import MazCard from '@components/MazCard.vue'

test('components/MazCard.vue', () => {
  expect(MazCard).toBeTruthy()

  const wrapper = mount(MazCard, {
    props: {},
  })

  expect(wrapper.html()).toMatchSnapshot()
})
