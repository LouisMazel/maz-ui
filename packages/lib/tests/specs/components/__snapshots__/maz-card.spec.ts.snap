// Vitest Snapshot v1

exports[`components/MazCard.vue 1`] = `
"<div class=\\"m-card maz-elevation maz-rounded-xl\\" data-v-49a838f4=\\"\\">
  <!--v-if-->
  <div class=\\"m-card__wrapper m-card__wrapper--column\\" data-v-49a838f4=\\"\\">
    <!--v-if-->
    <div class=\\"maz-min-w-0 maz-flex-1\\" data-v-49a838f4=\\"\\">
      <div data-v-49a838f4=\\"\\">
        <div class=\\"maz-p-4 m-card__content__wrapper\\" data-v-49a838f4=\\"\\">
          <!--v-if-->
          <!--v-if-->
          <!--v-if-->
        </div>
      </div>
    </div>
  </div>
  <!--v-if-->
  <!--v-if-->
</div>"
`;
